crates/compiler/serialize/src/bytes.rs

```diff
@@ -0,0 +1,381 @@
+where
+    T: Copy,
+    Container: From<Vec<T>>,
+{
+    let (serialized_slices, offset) = deserialize_slice::<VecSlice<T>>(bytes, length, offset);
+
+    let (vars_slice, offset) = {
+        let total_items = serialized_slices.iter().map(|s| s.len()).sum();
+        deserialize_slice::<T>(bytes, total_items, offset)
+    };
+
+    let mut slice_of_slices = Vec::with_capacity(length);
+    for slice in serialized_slices {
+        let deserialized_slice = &vars_slice[slice.indices()];
+        slice_of_slices.push(deserialized_slice.to_vec().into())
+    }
+
+    (slice_of_slices, offset)
+}
+
+pub fn serialize_map<K: Clone, V: Clone, W: Write>(
+    map: &MutMap<K, V>,
+    ser_keys: fn(&[K], &mut W, usize) -> io::Result<usize>,
+    ser_values: fn(&[V], &mut W, usize) -> io::Result<usize>,
+    writer: &mut W,
+    written: usize,
+) -> io::Result<usize> {
+    let keys = map.keys().cloned().collect::<Vec<_>>();
+    let values = map.values().cloned().collect::<Vec<_>>();
+
+    let written = ser_keys(keys.as_slice(), writer, written)?;
+    let written = ser_values(values.as_slice(), writer, written)?;
+
+    Ok(written)
+}
+
+#[allow(clippy::type_complexity)]
+pub fn deserialize_map<K, V>(
+    bytes: &[u8],
+    deser_keys: fn(&[u8], usize, usize) -> (Vec<K>, usize),
+    deser_values: fn(&[u8], usize, usize) -> (Vec<V>, usize),
+    length: usize,
+    offset: usize,
+) -> (MutMap<K, V>, usize)
+where
+    K: Clone + std::hash::Hash + Eq,
+    V: Clone,
+{
+    let (keys, offset) = deser_keys(bytes, length, offset);
+    let (values, offset) = deser_values(bytes, length, offset);
+
+    (
+        MutMap::from_iter((keys.iter().cloned()).zip(values.iter().cloned())),
+        offset,
+    )
+}
+
+pub fn serialize_vec_map<K, V, W: Write>(
+    map: &VecMap<K, V>,
+    ser_keys: fn(&[K], &mut W, usize) -> io::Result<usize>,
+    ser_values: fn(&[V], &mut W, usize) -> io::Result<usize>,
+    writer: &mut W,
+    written: usize,
+) -> io::Result<usize> {
+    let (keys, values) = map.unzip_slices();
+
+    let written = ser_keys(keys, writer, written)?;
+    let written = ser_values(values, writer, written)?;
+
+    Ok(written)
+}
+
+#[allow(clippy::type_complexity)]
+pub fn deserialize_vec_map<K, V>(
+    bytes: &[u8],
+    deser_keys: fn(&[u8], usize, usize) -> (Vec<K>, usize),
+    deser_values: fn(&[u8], usize, usize) -> (Vec<V>, usize),
+    length: usize,
+    offset: usize,
+) -> (VecMap<K, V>, usize)
+where
+    K: PartialEq,
+{
+    let (keys, offset) = deser_keys(bytes, length, offset);
+    let (values, offset) = deser_values(bytes, length, offset);
+
+    (unsafe { VecMap::zip(keys, values) }, offset)
+}
+
+unsafe fn slice_as_bytes<T>(slice: &[T]) -> &[u8] {
+    let ptr = slice.as_ptr();
+    let byte_length = std::mem::size_of::<T>() * slice.len();
+
+    std::slice::from_raw_parts(ptr as *const u8, byte_length)
+}
+
+fn round_to_multiple_of(value: usize, base: usize) -> usize {
+    (value + (base - 1)) / base * base
+}
```

Comment:
round up or down? if up, there is a rust function for that that will hopefully get stabilized soon. Until then, we can use
```rust
// check on https://github.com/rust-lang/rust/issues/88581 some time in the future
pub const fn next_multiple_of(lhs: usize, rhs: usize) -> usize {
    match lhs % rhs {
        0 => lhs,
        r => lhs + (rhs - r),
    }
}
```

Reply:
Rolled up into #4290